File: users-service/src/main/java/net/plshark/auth/service/DefaultTokenVerifier.java
package net.plshark.auth.service;

import java.util.Objects;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.exceptions.JWTVerificationException;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.auth0.jwt.interfaces.Payload;
import org.springframework.security.authentication.BadCredentialsException;
import reactor.core.publisher.Mono;

/**
 * Verifies JWTs
 */
public class DefaultTokenVerifier implements TokenVerifier{

    private final JWTVerifier verifier;

    /**
     * Create a new instance
     * @param verifier the verifier to use to verify and decode JWTs
     */
    public DefaultTokenVerifier(JWTVerifier verifier) {
        this.verifier = Objects.requireNonNull(verifier);
    }

    @Override
    public Mono<String> verifyToken(String token) {
        return decodeToken(token)
                .map(Payload::getSubject);
    }

    @Override
    public Mono<String> verifyRefreshToken(String token) {
        return decodeToken(token)
                .filter(validToken -> Boolean.TRUE == validToken.getClaim(AuthService.REFRESH_CLAIM).asBoolean())
                .switchIfEmpty(Mono.defer(() -> Mono.error(new BadCredentialsException("Token is not a refresh token"))))
                .map(Payload::getSubject);
    }

    /**
     * Validate and decode a token
     * @param token the token
     * @return the decoded token if successful or a BadCredentialsException if the token is invalid
     */
    private Mono<DecodedJWT> decodeToken(String token) {
        return Mono.defer(() -> {
            try {
                return Mono.just(verifier.verify(token));
            } catch (JWTVerificationException e) {
                throw new BadCredentialsException("Invalid credentials", e);
            }
        });
    }
}

File: user-error-server/src/main/kotlin/net/plshark/usererror/server/authentication/token/DefaultTokenVerifier.kt
package net.plshark.usererror.server.authentication.token

import com.auth0.jwt.JWTVerifier
import com.auth0.jwt.exceptions.JWTVerificationException
import com.auth0.jwt.interfaces.DecodedJWT
import net.plshark.usererror.authentication.token.UserErrorClaim
import net.plshark.usererror.authorization.UserAuthorities
import org.springframework.security.authentication.BadCredentialsException
import java.util.Optional

/**
 * Verifies JWTs
 */
class DefaultTokenVerifier(private val verifier: JWTVerifier) : TokenVerifier {

    override fun verifyToken(token: String): UserAuthorities {
        val jwt = decodeToken(token)
        return UserAuthorities(
            username = jwt.subject,
            authorities = parseAuthorities(jwt)
        )
    }

    override fun verifyRefreshToken(token: String): String {
        val jwt = decodeToken(token)
        val refresh = jwt.getClaim(UserErrorClaim.REFRESH).asBoolean()
        if (refresh == null || !refresh) throw BadCredentialsException("Token is not a refresh token")
        return jwt.subject
    }

    /**
     * Validate and decode a token
     * @param token the token
     * @return the decoded token if successful or a BadCredentialsException if the token is invalid
     */
    private fun decodeToken(token: String): DecodedJWT {
        try {
            return verifier.verify(token)
        } catch (e: JWTVerificationException) {
            throw BadCredentialsException("Invalid credentials", e)
        }
    }

    /**
     * Parse a list of authorities from a decoded JWT
     * @param jwt the decoded JWT
     * @return the list of authorities from the token
     */
    private fun parseAuthorities(jwt: DecodedJWT): Set<String> {
        return Optional.ofNullable(jwt.getClaim(UserErrorClaim.AUTHORITIES).asList(String::class.java))
            .map { list -> list.toSet() }
            .orElse(emptySet())
    }
}
